File: Code/Java/Longest_Common_Subsequence.java
/*LCS Problem Statement: Given two sequences, find the length of longest subsequence present in both of them. 
 * A subsequence is a sequence that appears in the same relative order, but not necessarily contiguous. 
 * For example, “abc”, “abg”, “bdf”, “aeg”, ‘”acefg”, .. etc are subsequences of “abcdefg”.*/

import java.util.Scanner;
public class Longest_Common_Subsequence {
	public static int LengthOfLCS(String s, String t) {
		     int m = s.length();
		     int n = t.length();
		     int storage[][] = new int[m + 1][n + 1];
		     // Base case for first row
		     for (int i = 0; i <= m; i++) {
			         storage[i][0] = 0;
		     }
         // Base case for first column
		     for (int i = 0; i <= n; i++) {
			          storage[0][i] = 0;
		     }
         for (int i = 1; i <= m; i++) {
			        for (int j = 1; j <= n; j++) {
				           if (s.charAt(m - i) == t.charAt(n - j)) {
                         storage[i][j] = 1 + storage[i - 1][j - 1];
                   } else {
                          storage[i][j] = Math.max(storage[i][j - 1], storage[i - 1][j]);
				            }
			          }
	      	}
		      return storage[m][n];
	}
	public static String printTheLCS(String s, String t) {
		     int[][]Storage = new int[s.length() + 1][t.length() + 1];
         int m = s.length();
         int n = t.length();
		     // Following steps build storage[m+1][n+1] in bottom up fashion
		           // storage[i][j] contains the maximum length of LCS of s[0..i-1] and t[0..j-1]
		            for (int i = 0; i <= s.length(); i++) {
			               for (int j = 0; j <= t.length(); j++) {
			                   	   // Base case 
				                     if (i == 0 || j == 0) {
				                        	Storage[i][j] = 0;
				                     } else if (s.charAt(i - 1) == t.charAt(j - 1)) {
					                        Storage[i][j] = Storage[i - 1][j - 1] + 1;
				                     } else {
					                        Storage[i][j] = Math.max(Storage[i - 1][j], Storage[i][j - 1]);
				                     }	
			                }
	            	}
           // Following code is used to print LCS
		       int index = Storage[s.length()][t.length()];
		       int temp = index;
        
           // Create a character array to store the lcs string
		           char[] lcs = new char[index + 1];
		           lcs[index] = '\u0000'; // Set the terminating character
        
           // Start from the right-most-bottom-most corner and
		       // one by one store characters in lcs[]
		           int i = m;
		           int j = n;
		           while (i > 0 && j > 0) {
			               // If current character in s and t are same, then current character is part of LCS
			                  if (s.charAt(i - 1) == t.charAt(j - 1)) {
			                         // Put current character in result
				                         lcs[index - 1] = s.charAt(i - 1);

				                       // reduce values of i, j and index
			                           	i--;
				                          j--;
				                          index--;
			                    }     // If not same, then find the larger of two and go in the direction of larger value
			                    else if (Storage[i - 1][j] > Storage[i][j - 1]) {
				                           i--;
		                      } else {
			                             j--;
			                    }	
		           }
		           String ans = "";
		           for (int k = 0; k <= temp; k++) {
		               	ans += lcs[k];
		            }
		    return ans;
    } 
  
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		String str1 = sc.next();
		String str2 = sc.next();
		System.out.print("Longest Common Subsequence of " + str1 + " and " + str2 + " is " + printTheLCS(str1,str2) + 
				         "of length "+ LengthOfLCS(str1,str2));
		sc.close();
	}
}
/*
Sample Input  1: Str1 : "ABCDGH"  Str2 : "AEDFHR"
Sample Output 1: Longest Common Subsequence of “ABCDGH” and “AEDFHR” is “ADH” of length 3.
Sample Input  2: Str1 : "AGGTAB"  Str2 : "GXTXAYB"
Sample Output 2: Longest Common Subsequence of “AGGTAB” and “GXTXAYB” is “GTAB” of length 4.

Time Complexity : O(M*N)
Auxiliary Space : O(N)
*/

File: Code/Python/longest_common_subsequence.py
###################################################################
# LCS Algorithm

# Given two X=<x1,x2,....xn> and Y=<y1,y2,....,yn>, We have to find a common subsequence whose length is maximum.

# Subsequence need not be consecutive, but must be in order.

# Subsequence in sequence that appears in same relative order,but not neccessarily contiguous.

# More than 1 LCS can be present between two sequence , from those we have to find the subsequence of maximum length.

######################################################################

# Program Using Dynamic Programming Approach

X = input("Enter the first String -> ")
Y = input("Enter the second String -> ")
# Length of the two strings
m = len(X)
n = len(Y)
L = [[0 for i in range(n + 1)] for i in range(m + 1)]

# Algorithm for LCS
def lcs(X, Y):
    i = 0
    j = 0

    # Base Conditon For rows
    for i in range(1, m + 1):
        L[i][0] = 0
    # Base Conditon For columns
    for j in range(0, n + 1):
        L[0][j] = 0

    # Building L[m + 1][n + 1] in bottom up fashion
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if X[i - 1] == Y[j - 1]:
                L[i][j] = L[i - 1][j - 1] + 1
            elif L[i - 1][j] >= L[i][j - 1]:
                L[i][j] = L[i - 1][j]
            else:
                L[i][j] = L[i][j - 1]


def lcs_display(X, Y):
    lcs(X, Y)

    # Length of LCS
    length = L[m][n]

    # Declaring a list of string to store the resultant LCS
    index = L[m][n]
    disp_lcs = [""] * (index + 1)
    disp_lcs[index] = ""

    i = m
    j = n

    # Getting a charcter one by one from the right bottom-up approach
    while i > 0 and j > 0:
        # If ith and jth charcter at the same position are same,then include the character
        if X[i - 1] == Y[j - 1]:
            disp_lcs[index - 1] = X[i - 1]
            i -= 1
            j -= 1
            index -= 1
        # If not equal then follow the path of the larger value
        elif L[i - 1][j] > L[i][j - 1]:
            i -= 1
        else:
            j -= 1

    # Display the LCS and its length
    print(
        f'Longest Common Subsequence of {X} and {Y} is {"".join(disp_lcs)} which is of length {length}'
    )


lcs_display(X, Y)

# Sample Input 1
# X="springtime" , Y="printing"

# Sample Output 1
# Longest Common Subsequence of springtime and printing is printi which is of length 6

# Sample Input
# X="abcdefghi" , Y= "ecdgi"

# Sample Output 2
# Longest Common Subsequence of abcdefghi and ecdgi is cdgi which is of length 4

# Time Complexity : O(m*n)
# Auxiliary Space : O(n)